// Copyright (C) 2020-2023 Intel Corporation
// Licensed subject to the terms of the separately executed evaluation license agreement between Intel Corporation and you.

syntax = "proto3";

// This is only needed to initializing the model weights on the aggregator
message ModelProto {
  repeated NamedTensor tensors = 1;
}

message NamedTensor {
  string name = 1;
  int32 round_number = 2;
  bool lossless	= 3;
  bool report = 4;
  repeated string tags = 5;
  repeated MetadataProto transformer_metadata = 6;
  bytes data_bytes = 7;
}

message MetadataProto {
  map<int32, float> int_to_float = 1;
  repeated int32 int_list = 2;
  repeated bool bool_list = 3;
}

// handles large size data
message DataStream {
  uint32 size = 1; // size, in bytes, of the data sent in npbytes
  bytes npbytes = 2; // actual data
}

message CollaboratorDescription {
  string name = 1;
  string status = 2;
  float progress = 3;
  uint32 round = 4;
  string current_task = 5;
  string next_task = 6;
}

// The same with director.proto
message TaskDescription {
  string name = 1;
  string description = 2;
}

message DownloadStatus {
  string name = 1;
  string status = 2;
}

message DownloadStatuses {
  repeated DownloadStatus models = 1;
  repeated DownloadStatus logs = 2;
}

message ExperimentDescription {
  string name = 1;
  string status = 2;
  float progress = 3;
  uint32 total_rounds = 4;
  uint32 current_round = 5;
  DownloadStatuses download_statuses = 6;
  repeated CollaboratorDescription collaborators = 7;
  repeated TaskDescription tasks = 8;
}
